### DISCUSSION TITLE
Help with output of program

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#20450] [Q&A Modules: Solid mechanics] [answered]
Hi Moose experts or users,

Recently, I develop a custom class inherited with the base class of 'CrystalPlasticityStressUpdateBase' to study thermal creep behavior based on one creep model ('A physics-based crystallographic modeling framework for describing the thermal creep behavior of Fe-Cr alloys'). "Creep testing is conducted using a tensile specimen to which a constant stress is applied at a constant temperature, often by the simple method of suspending weights from it." According to this test method, the boundary conditions in my test file with .i extension are set as follows: a constant tensile pressure is applied on the top surface of a single element; a constant temperature is applied on the top and bottom surface. The left and back are fixed. The program can be successfully performed. But the output of stress and strain are zero. Could you share any thoughts or suggestions with me to solve it?
<img width="582" alt="Screen Shot 2022-03-01 at 5 25 54 PM" src="https://user-images.githubusercontent.com/74226455/156266493-fd1ceec4-c812-430e-bdf9-95f1b850f9e1.png">

**Comment by sapitts** [comment, marked as answer]
Without seeing your code and input file, I would guess that you have not fully connected the constitutive model to the stress and strain measures calculated in `ComputeMultipleCrystalPlasticityStress`. Have you made sure to override all of the virtual functions in `CrystalPlasticityStressUpdateBase`, like `calculateSlipRate` and `calculateConstitutiveSlipDerivative`?

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for your response. Could you share more with me about how to make sure override all of the virtual functions in 'CrystalPlasticityStressUpdateBase', like 'calculateSlipRate' and 'calculateConstitutiveSlipDerivative' you mentioned? Thank you.

**Reply by sapitts** [reply]
In the base class header file, look for method definitions that are equal to zero, e.g.
```
  /**
   * This virtual method is called to calculate the slip system slip
   * increment based on the constitutive model defined in the child class.
   * This method must be overwritten in the child class.
   */
  virtual bool calculateSlipRate() = 0;
  ```

Your inheriting classes should then provide a (non zero) definition for these methods, e.g. from [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.h](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.h#L37)
and then define the appropriate constitutive model in the source file, e.g. [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C](https://github.com/idaholab/moose/blob/8800a0bf87e3822e9d57d39c7ec133caa7b9e57d/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L334)

Once you have made sure that you have overrides for all of the methods, I'd recommend printing out the material properties like `slip_increment` and the `plastic_deformation_gradient`.

Crystal plasticity models are difficult to debug and often take a long time to get all the bugs worked out. I often work on single element models, with A LOT of print statements, to debug and verify my crystal plasticity models

**Reply by xchengood** [reply]
Appreciate your information. I will carefully check if all of the methods have been overridden and try to print out those variables. Thank you again.

**Reply by xchengood** [reply]
Hi @sapitts. Sorry for bothering you again. Now I know how to use the lldb debugger, how to set a breakpoint in a certain location, and how to print out the content of a variable. Could you use 'CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C' for example to resolve the following questions.

1. I have no experience on how to set an appropriate breakpoint in somewhere the source file. I think it is important if I want to check the value of the variable.
2. Some variables are defined by vector (e.g. _slip_increment). How can I display the content of this vector using lldb? What I did was to specify one, for example, _slip_increment[_qp][1], and then print out one by one. Do you have good advice?
3. I can only print out global variables but have no idea how to display the local variables defined in a function.

Thank you.

**Reply by sapitts** [reply]
Hi @xchengood, I'm glad to help as I can! I must admit, I generally don't use anything as sophisticated as lldb. Instead, I usually have a bunch of old fashioned `std::cout << ` statements all over my code, and I try to have these print statements for each qp and each slip system.

1. Maybe @dewenyushu or @dschwen have some suggestions on setting breakpoints for looking at the values of variables
2. When I print out the value of variables with `std::cout` I try to be inside of a loop over the slip systems to use the loop to index the vector. For example, in the `calculateSlipRate` method of [CrystalPlasticityTwinningKalidindiUpdate](https://github.com/idaholab/moose/blob/8800a0bf87e3822e9d57d39c7ec133caa7b9e57d/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C#L147), I used the following to print out the value of the slip increment, lines 150-164 plus a few extra print statements:
```
for (const auto i : make_range(_number_slip_systems))
    total_twin_volume_fraction += _twin_volume_fraction[_qp][i];

  if (total_twin_volume_fraction < _limit_twin_volume_fraction)
  {
    for (const auto i : make_range(_number_slip_systems))
    {
      if (_tau[_qp][i] > 0.0)
      {
        const Real driving_force = (_tau[_qp][i] / _slip_resistance[_qp][i]);
        _slip_increment[_qp][i] = std::pow(driving_force, (1.0 / _rate_sensitivity_exponent)) *
                                  _reference_strain_rate * _substep_dt;
      }
      else // twin propagation is directional
        _slip_increment[_qp][i] = 0.0;

   std::cout << " On qp " << _qp << " and twin system number " << i << ":\n";
   std::cout << "   the value of the slip increment is " << _slip_increment[_qp][i] << "\n";
```
At least this way you don't have to explicitly call each entry in the vector (which could be a large pain if you have the 24 or 48 slip systems!)
3. Using the unsophisticated `std::cout` method listed above will also work for local variables!

Good luck!

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for your helpful suggestions very much. I forget I can use this simple code to show the content of any variables. I will try more and check the variables. Appreciate.

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for giving me helpful instructions on debugging. Now, I got the following errors. Could you look through and share any advice with me to fix it?
[errors.txt](https://github.com/idaholab/moose/files/8189150/errors.txt)
![Screen Shot 2022-03-04 at 4 48 38 PM](https://user-images.githubusercontent.com/74226455/156852793-6088ebb0-b5df-444a-bc4d-4908155b108a.png)

**Reply by GiudGiud** [reply]
@xchengood the time step is being cut down, eventually leading to an error for minimum time step size, because of the constitutive failure in ComputeMultipleCrystalPlasticityStress. 

Namely this routine is failing to reach convergence:
```ComputeMultipleCrystalPlasticityStress::updateStress```

can you check that your initial conditions are physically acceptable?
I recommend you place a few print statement in that routine to see if something is off.
Alternatively, use a debugger, place a breakpoint on this routine and step through it, printing variables regularly

**Reply by sapitts** [reply]
Hi @xchengood, in addition to the initial condition check @GiudGiud recommended, I'd also recommend looking at your dislocation density values, slip (and related) system resistance, and the applied shear stress.  Try printing out your constitutive model quantities to ensure that those values are inline with your expectations (e.g. only positive dislocation densities and slip system resistances)

**Reply by xchengood** [reply]
@GiudGiud and @sapitts. I appreciate your information. You help me a lot indeed. I will do more attempts according to your advice. Thank you.

**Reply by xchengood** [reply]
Hi @sapitts. I know I have to override two functions, **calculateSlipRate()** and **calculateConstitutiveSlipDerivative(std::vector<Real> & dslip_dtau)**. According to the CP model I used shown below, 
<img width="450" alt="Screen Shot 2022-03-08 at 10 01 05 AM" src="https://user-images.githubusercontent.com/74226455/157277739-75930a28-bcc0-4738-b262-7130c874f57c.png">

The first function is expressed in Moose as follows:
`_slip_increment[_qp][i] = _forest_dislocation_density[_qp][i] * _burgers_vector *
                                _dislocation_velocity[_qp][i] * std::copysign(1.0, _tau[_qp][i]);`

But for the second function, I am not sure if it is correct.
`dslip_dtau[i] = _forest_dislocation_density[_qp][i] * _burgers_vector *
                        _dislocation_velocity[_qp][i] * _substep_dt;`

Seems not correct. Thank you.

**Reply by sapitts** [reply]
The first function is computing a slip increment, so you probably should multiply by the `_substep_dt` to convert from the rate form given in the paper to the increment form in the code.

The second function is looking for the derivative of the slip increment with respect to the applied shear stress. You should derive this expression and implement that derivative. Likely in your case the slip increment depends on the applied shear stress through the velocity term

**Reply by xchengood** [reply]
@sapitts, Thank you. Good advice. I will try to derive that expression.
